File: day-19a/src/main.py
from __future__ import annotations

from collections.abc import Callable, Mapping, Sequence
from dataclasses import dataclass
from enum import Enum
from typing import NamedTuple, Self


class DecisionKind(Enum):
    ACCEPT = "A"
    REJECT = "R"
    OTHER_WORKFLOW = None


class Decision(NamedTuple):
    kind: DecisionKind
    data: str | None = None

    @classmethod
    def from_string(cls, s: str) -> Self:
        match s:
            case "A" | "R":
                return cls(DecisionKind(s))
            case _:
                return cls(DecisionKind.OTHER_WORKFLOW, data=s)


@dataclass(slots=True)
class Part:
    x: int
    m: int
    a: int
    s: int

    def score(self) -> int:
        return self.x + self.m + self.a + self.s

    @classmethod
    def from_string(cls, s: str) -> Self:
        part = cls(0, 0, 0, 0)
        for section in s[1:-1].split(","):
            attr, _, value_str = section.partition("=")
            setattr(part, attr, int(value_str))
        return part


def rule_from_string(s: str) -> Callable[[Part], Decision | None]:
    match list(s):
        case ['x' | 'm' | 'a' | 's' as attr, '>' | '<' as cmp, *rest]:
            digits, _, outcome = "".join(rest).partition(":")
            value = int(digits)
            decision = Decision.from_string(outcome)
            match cmp:
                case ">":
                    return lambda p: decision if getattr(p, attr) > value else None
                case "<":
                    return lambda p: decision if getattr(p, attr) < value else None
        case _:
            decision = Decision.from_string(s)
            return lambda _: decision


@dataclass(slots=True, frozen=True, kw_only=True)
class Workflow:
    name: str
    rules: Sequence[Callable[[Part], Decision | None]]

    @classmethod
    def from_string(cls, s: str) -> Self:
        name, _, rule_strings = s.strip()[:-1].partition("{")
        rules = tuple(
            rule_from_string(rule_string) for rule_string in rule_strings.split(",")
        )
        return cls(name=name, rules=rules)

    def process(self, part: Part) -> Decision:
        for rule in self.rules:
            ruling = rule(part)
            if ruling is not None:
                return ruling
        assert False, (
            "At least one rule in self.rules should have returned a `Decision`!"
        )

    def __str__(self) -> str:
        return f"Workflow({self.name!r}, <{len(self.rules)} rules>)"


@dataclass(slots=True, frozen=True, kw_only=True)
class PuzzleInput:
    workflow_map: Mapping[str, Workflow]
    parts: Sequence[Part]

    @classmethod
    def from_string(cls, s: str) -> Self:
        workflow_strings, _, part_strings = s.partition("\n\n")
        workflows = (
            Workflow.from_string(line) for line in workflow_strings.split("\n")
        )
        workflow_map = {w.name: w for w in workflows}
        parts = tuple(Part.from_string(line) for line in part_strings.split("\n"))
        return cls(workflow_map=workflow_map, parts=parts)


def solve(filename: str) -> int:
    with open(filename) as f:
        puzzle_input = PuzzleInput.from_string(f.read())
    answer = 0
    for part in puzzle_input.parts:
        outcome = Decision.from_string("in")
        while True:
            match outcome.kind:
                case DecisionKind.ACCEPT:
                    answer += part.score()
                    break
                case DecisionKind.REJECT:
                    break
                case DecisionKind.OTHER_WORKFLOW:
                    assert outcome.data is not None
                    workflow = puzzle_input.workflow_map[outcome.data]
                    outcome = workflow.process(part)
    return answer


def main() -> None:
    print(solve("input.txt"))


if __name__ == "__main__":
    main()

File: day-19a/src/main.rs
use std::collections::HashMap;
use std::fmt::Display;
use std::fs::read_to_string;
use std::str::FromStr;

use anyhow::{bail, Context, Error, Result};

#[derive(Debug)]
enum Decision {
    Accept,
    Reject,
    OtherWorkflow(String),
}

impl From<&str> for Decision {
    fn from(s: &str) -> Self {
        match s {
            "A" => Self::Accept,
            "R" => Self::Reject,
            _ => Self::OtherWorkflow(s.to_string()),
        }
    }
}

#[derive(Debug, Clone, Copy)]
struct Part {
    x: u32,
    m: u32,
    a: u32,
    s: u32,
}

impl Part {
    fn score(&self) -> u32 {
        self.x + self.m + self.a + self.s
    }
}

impl FromStr for Part {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self> {
        let mut data = HashMap::new();
        let sections = s[1..(s.len() - 1)].split(',');
        for section in sections {
            let split_section = Vec::from_iter(section.split('='));
            let rating = u32::from_str(split_section[split_section.len() - 1])?;
            data.insert(split_section[0], rating);
        }
        Ok(Self {
            x: data["x"],
            m: data["m"],
            a: data["a"],
            s: data["s"],
        })
    }
}

#[derive(Debug, Clone, Copy)]
enum Compare {
    Lt,
    Gt,
    NoOp,
}

impl TryFrom<&char> for Compare {
    type Error = anyhow::Error;

    fn try_from(value: &char) -> Result<Self> {
        match value {
            '>' => Ok(Self::Gt),
            '<' => Ok(Self::Lt),
            _ => bail!("Don't know how to create a `Compare` variant from {value}"),
        }
    }
}

#[derive(Debug, Clone, Copy)]
enum Attr {
    X,
    M,
    A,
    S,
}

impl TryFrom<&char> for Attr {
    type Error = anyhow::Error;

    fn try_from(value: &char) -> Result<Self> {
        match value {
            'x' => Ok(Attr::X),
            'm' => Ok(Attr::M),
            'a' => Ok(Attr::A),
            's' => Ok(Attr::S),
            _ => bail!("Don't know how to create an `Attr` from {value}"),
        }
    }
}

struct Rule {
    attr: Option<Attr>,
    cmp: Compare,
    value: u32,
    outcome: Decision,
}

impl Rule {
    fn new(attr: Attr, cmp: Compare, value: u32, outcome: Decision) -> Self {
        assert!(!matches!(cmp, Compare::NoOp));
        Rule {
            attr: Some(attr),
            cmp,
            value,
            outcome,
        }
    }

    fn noop(outcome: Decision) -> Self {
        Rule {
            attr: None,
            cmp: Compare::NoOp,
            value: 0,
            outcome,
        }
    }

    fn process(&self, part: &Part) -> Option<Decision> {
        let Rule {
            attr,
            cmp,
            value,
            outcome,
        } = self;
        let inner: Box<dyn Fn(&Part) -> bool> = match (attr, cmp) {
            (Some(Attr::X), Compare::Gt) => Box::new(|p: &Part| p.x > *value),
            (Some(Attr::X), Compare::Lt) => Box::new(|p: &Part| p.x < *value),
            (Some(Attr::M), Compare::Gt) => Box::new(|p: &Part| p.m > *value),
            (Some(Attr::M), Compare::Lt) => Box::new(|p: &Part| p.m < *value),
            (Some(Attr::A), Compare::Gt) => Box::new(|p: &Part| p.a > *value),
            (Some(Attr::A), Compare::Lt) => Box::new(|p: &Part| p.a < *value),
            (Some(Attr::S), Compare::Gt) => Box::new(|p: &Part| p.s > *value),
            (Some(Attr::S), Compare::Lt) => Box::new(|p: &Part| p.s < *value),
            (None, Compare::NoOp) => Box::new(|_: &Part| true),
            _ => unreachable!("The combination of {attr:?} and {cmp:?} should be impossible!",),
        };
        if inner.as_ref()(part) {
            let outcome = match outcome {
                Decision::Accept => Decision::Accept,
                Decision::Reject => Decision::Reject,
                Decision::OtherWorkflow(s) => Decision::OtherWorkflow(s.to_owned()),
            };
            Some(outcome)
        } else {
            None
        }
    }
}

impl FromStr for Rule {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self> {
        match &s.chars().collect::<Vec<char>>()[..] {
            [attr @ ('x' | 'm' | 'a' | 's'), cmp @ ('>' | '<'), rest @ ..] => {
                let attr = Attr::try_from(attr)?;
                let cmp = Compare::try_from(cmp)?;
                let rest = String::from_iter(rest);
                let [digits, outcome] = rest.split(':').collect::<Vec<_>>()[..] else {
                    bail!("Don't know how to create a Rule from {s}")
                };
                let value = u32::from_str(digits)?;
                let outcome = Decision::from(outcome);
                Ok(Rule::new(attr, cmp, value, outcome))
            }
            chars @ [..] => {
                let outcome = Decision::from(String::from_iter(chars).as_str());
                Ok(Rule::noop(outcome))
            }
        }
    }
}

struct Workflow {
    name: String,
    rules: Vec<Rule>,
}

impl FromStr for Workflow {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self> {
        let s = s.trim();
        let s = &s[..(s.len() - 1)];
        let [name, rule_strings] = s.split('{').collect::<Vec<_>>()[..] else {
            bail!("Unexpected number of braces in {s}")
        };
        let rules = rule_strings
            .split(',')
            .map(Rule::from_str)
            .collect::<Result<_>>()?;
        Ok(Workflow {
            name: name.to_string(),
            rules,
        })
    }
}

impl Workflow {
    fn process(&self, part: Part) -> Decision {
        for rule in &self.rules {
            if let Some(decision) = rule.process(&part) {
                return decision;
            }
        }
        unreachable!("At least one rule in self.rules should have returned a `Decision` variant!")
    }
}

impl Display for Workflow {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let Workflow { name, rules } = self;
        write!(f, "Workflow(\"{name}\", <{} rules>)", rules.len())
    }
}

struct PuzzleInput {
    workflow_map: HashMap<String, Workflow>,
    parts: Vec<Part>,
}

impl FromStr for PuzzleInput {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self> {
        let string = s.replace("\r\n", "\n");
        let [workflow_strings, part_strings] = string.split("\n\n").collect::<Vec<&str>>()[..]
        else {
            bail!("Unexpectedly found more than one double-linebreak in the puzzle input!")
        };
        let workflows = workflow_strings
            .lines()
            .map(|line| line.parse())
            .collect::<Result<Vec<Workflow>>>()?;
        let mut workflow_map = HashMap::new();
        for workflow in workflows {
            workflow_map.insert(workflow.name.to_owned(), workflow);
        }
        let parts = part_strings
            .lines()
            .map(|line| line.parse())
            .collect::<Result<Vec<Part>>>()?;
        Ok(PuzzleInput {
            workflow_map,
            parts,
        })
    }
}

fn parse_input(filename: &str) -> Result<PuzzleInput> {
    let input_string = read_to_string(filename)
        .with_context(|| format!("Expected {filename} to exist as a file!"))?;
    PuzzleInput::from_str(&input_string)
}

fn solve(filename: &str) -> u32 {
    let input = parse_input(filename).unwrap();
    let mut answer = 0;
    for part in input.parts {
        let mut outcome = Decision::OtherWorkflow("in".to_string());
        loop {
            match outcome {
                Decision::Accept => {
                    answer += part.score();
                    break;
                }
                Decision::Reject => break,
                Decision::OtherWorkflow(ref s) => outcome = input.workflow_map[s].process(part),
            }
        }
    }
    answer
}

fn main() {
    println!("{}", solve("input.txt"));
}
